### ISSUE TITLE
Possible conflict between role tooltip and WCAG 2.1 Success Criterion 1.4.13 Content on Hover or Focus

### ISSUE THREAD

**Issue by Decrepidos** [w3c/aria-practices#811]
### Issue 1:

**ARIA:**
[https://www.w3.org/TR/wai-aria-practices-1.1/#tooltip](url)

If the tooltip is invoked when the **trigger element** receives focus, then it is dismissed when **it** no longer has focus (onBlur). If the tooltip is invoked with mouseIn, then it is dismissed with on mouseOut.

**WCAG 2.1**
[https://www.w3.org/TR/WCAG21/#content-on-hover-or-focus](url)

If pointer hover can trigger the additional content, then the pointer can be **moved over the additional content without the additional content disappearing**.
The **additional content remains visible until the hover or focus trigger is removed**, the user dismisses it, or its information is no longer valid.

In the ARIA spec it is the **triggering element (link/button)** only that controls the show/hide of the tooltip content (additional content). This conflicts with the WCAG spec where focus needs to **encompass the trigger and additional content**.

### Issue 2:

[https://www.w3.org/WAI/WCAG21/Understanding/content-on-hover-or-focus.html](url)

As I understand it role=tooltip does not allow active embedded elements, however, the WCAG 2.1 spec stats: 

Provide a mechanism to easily dismiss the additional content, such as by pressing Escape or **selecting a close button**.

**Comment by carmacleod** [comment]
@jlukosky 
> Since tooltips don't receive focus would your proposal be only for tooltip dialogs or are you proposing to make tooltips focusable?

Good question, and I don't have a good answer.  :)

When I saw the [meeting-bot](https://github.com/w3c/aria-practices/issues/811#issuecomment-410798024) comment containing tooltip discussion notes from the ARIA Working Group with "hh: need a shortcut to move focus to it", I simply wanted to make sure that if a shortcut is added to move focus to a tooltip (or tooltip dialog), then `SHIFT+F1` is seriously considered for that shortcut.

Tooltips (either title attribute or custom) have been a multi-pronged accessibility problem for so long that maybe it's time for some off-the-wall changes to their UI (such as perhaps allowing them to take focus under certain conditions). ;)

Anyhow, I did not mean to go off-topic for this issue, which is about WCAG SC 1.4.13.

**Comment by smhigley** [comment]
Based on a mix of discussion and opinion, here are the additions to the APG tooltip pattern I would suggest based on this issue. For more potentially controversial or opinion-based recommendations, I have added the prefix "(opinion)".

> Pointer Interaction:
> - The tooltip remains open when moving the mouse from the triggering element to the tooltip element.
> - (opinion) To benefit users with mobility impairments and to aid in easily moving the mouse between the trigger and the tooltip, there is a short delay between the mouseout event and the tooltip disappearing.
> - (opinion) Pointer users may dismiss the tooltip by selecting a close button, after which the tooltip will not reappear on hover. The close button supplements keyboard escape behavior, so does not need to receive keyboard focus.

That last one is probably the most touchy, so I'll expand on the reasons for it:

Escape is not always a feasible way of dismissing a tooltip when focus is not on the triggering element. If the tooltip exists within a dialog, or focus happens to be on an element like a combobox or menu button, pressing escape will be handled by the dialog/menu button/combobox. Escape will not only not dismiss the tooltip, but will also cause undesired side effects (e.g. you try to dismiss a tooltip but end up closing the entire dialog).

In addition to the practical limitations, it seems more user-friendly to provide a pointer-based method of dismissing tooltips. Since a close button would be solely for the benefit of pointer users, the simplest approach would likely be to remove it from the tab order and not treat it as interactive tooltip content that would need a separate keyboard shortcut.

If it makes more sense to move forward with the first, or first and second recommendations and table the third as out of scope, I'm definitely open to that too :).

**Comment by JAWS-test** [comment]
I have no problem with a close button for pointer users. But
* If ESC closes the entire dialog, this also applies to keyboard users who cannot reach and operate the button. This means that ESC should be implemented in such a way that it closes only the tooltip as long as the triggering element has the focus. To close the dialog, you would have to press ESC twice. If the triggering element has no more focus, the tooltip will close automatically and 1x ESC will close the dialog. 
* If ESC is pressed or the close button is activated, the tooltip should only be closed for this time and should not be deactivated permanently. In your example https://codepen.io/smhigley/pen/KjoerX, the tooltip is permanently closed and no longer displayed when I press ESC when focus the element again. If I press the close button, it will be displayed again.

**Comment by JAWS-test** [comment]
And your tooltips don't look nice in High Contrast Mode: missing frame and triangle becomes block
![tooltip](https://user-images.githubusercontent.com/52748763/64228718-ec601d80-cee7-11e9-908d-c6d4e197a2cc.jpg)

**Comment by smhigley** [comment]
@JAWS-test thanks for your comments, I'll do my best to address them in order:

- The concern with escape does not apply to keyboard users; it is solely an issue with using escape as the only means to dismiss a tooltip triggered by mouse hover. When a tooltip is triggered by a focus event, the triggering element can capture and handle escape key presses, thus preventing them from having undesired side effects. However, if the tooltip is triggered through hover, then the element that captures the escape key could be anything at all -- this is where the possibility of unintended consequences comes in.
- If a close button is used as a method of dismissal for pointer users, the dismissal would need to be at least semi-permanent to meet the spirit of 1.4.13's dismissible requirement. Imagine a ZoomText user navigating their cursor and field of view over the tooltip to dismiss it, only to re-center their view and have it pop up again. Perhaps one way around this would be to have the close button on the triggering control (I believe this was suggested some time ago in a different issue), but I haven't seen that pattern in the wild.
- I'm not sure why the codepen came up, but it is not at all related to this issue, or intended as a mock for the APG. It's explicitly not production-ready, and you've definitely hit on some good ways it could be improved. The code is open source, and the web could certainly use more accessible production-ready tooltips if you wanted to take it and run with it.

**Comment by JAWS-test** [comment]
@smhigley:

> The concern with escape does not apply to keyboard users

I understand your argument. However, it only applies if the application has been implemented correctly. I've already seen enough examples where the ESC key closed the entire pop-up, although the focus was on an element inside the pop-up that is closed with ESC.

> the dismissal would need to be at least semi-permanent to meet the spirit of 1.4.13

I agree with you. I had only seen the opposite behavior in your example: re-display for mouse users, but not for keyboard users. But the exciting question is how to do this semi-permanently, not permanently. Unfortunately I have no idea.

> I'm not sure why the codepen came up

I'm sorry. I knew your page and thought that your example should serve as a template for APG because it deals with exactly the same questions. That's why I had a closer look.

**Comment by JAWS-test** [comment]
The close button should not be inside the tooltip. It can be marked with aria-hidden=true, but the attribute will be ignored if it is referenced by aria-labelledby or aria-describedby. In JAWS, for example, the close button in the tooltip is output with Firefox and IE 11, but not with Chrome. See: https://github.com/FreedomScientific/VFO-standards-support/issues/251

**Comment by smhigley** [comment]
@JAWS-test 
> I understand your argument. However, it only applies if the application has been implemented correctly. [...]

The APG generally does not exclude or recommend against techniques solely because they have the potential to be misused. However, if you think that escape should never be the mechanism for dismissing tooltips, it might be better to raise that issue against WCAG, since escape is explicitly mentioned.

> The close button should not be inside the tooltip. It can be marked with aria-hidden=true, but the attribute will be ignored if it is referenced by aria-labelledby or aria-describedby.

Unless the close button itself is directly referenced by `aria-labelledby` or `aria-describedby`, it should not be included in the name or description: https://w3c.github.io/accname/#step2A. It would also be simple to point `aria-labelledby` or `aria-describedby` to an element containing only the tooltip text, while still keeping the button visually within the tooltip. There may be other reasons to consider different close button positions, but this doesn't seem to be one of them.

Hope that helps!

**Comment by jlukosky** [comment]
> * The tooltip remains open when moving the mouse from the triggering element to the tooltip element.

Would you be opposed to rewording this as 
"The tooltip remains open when moving the mouse between the triggering element and the tooltip element."?

I've seen at least one implementation that lets the user mouse over the tooltip but not back to the trigger without dismissing the tooltip.
